File: JavaScript/프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.js
function solution(enroll, referral, seller, amount) {
    var answer = [];
    const objCost = {};
    const objParent = {};
    for (const mem of enroll){
        objCost[mem] = 0;
    }
    for (const [idx, mem] of referral.entries()){
        if (mem[0] === '-') continue;
        objParent[enroll[idx]] = mem; // 부모 정해줌. 부모는 하나임.
    }
    for (let i = 0; i < seller.length; i++){
        let value = amount[i] * 100;
        let member = seller[i];
        while (member !== undefined && value > 0){
            let tenPercent = Math.floor(value * 0.1);
            objCost[member] += (value - tenPercent);
            member = objParent[member];
            value = tenPercent;
        }
    }
    for (let mem of enroll){
        answer.push(objCost[mem]);
    }
    return answer;
}

File: C++/프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.cpp
#include <string>
#include <vector>
#include <cmath>
#include <unordered_map>

using namespace std;

vector<int> solution(vector<string> enroll, vector<string> referral, vector<string> seller, vector<int> amount) {
    vector<int> answer(enroll.size(), 0);
    unordered_map<string, string> parent_info;
    unordered_map<string, int> pos_info;
    
    for (int i = 0; i < enroll.size(); i++){
        parent_info[enroll[i]] = referral[i];
        pos_info[enroll[i]] = i;
    }
    for (int i = 0; i < seller.size(); i++){
        int price = amount[i] * 100;
        string person = seller[i];
        
        while (person[0] != '-'){
            int keep = price * 0.1;
            
            answer[pos_info[person]] += price - keep;
            if (keep < 1) break;
            price = keep;
            person = parent_info[person];
        }
    }
    return answer;
}